### ISSUE TITLE
Full RELRO binary reported as Partial RELRO

### ISSUE THREAD

**Issue by biboon** [slimm609/checksec.sh#165]
## Issue

On FreeBSD 11.3, a binary compiled with clang and both RELRO and BIND_NOW is still detected as "Partial RELRO". However, the two options are supposed to be detected.

## Debug Report

```
root@freebsd:~ # checksec --debug-report
Error: Unknown option debug-report.
```

## Command run to produce the error

1. Get [FreeBSD virtual machine](https://download.freebsd.org/ftp/releases/VM-IMAGES/11.3-RELEASE/amd64/Latest/). I used the qcow2 image on virt-manager/QEMU.

2. Decompress the file.
   ```
   unxz -k FreeBSD-11.3-RELEASE-amd64.qcow2.xz
   ```

3. You may need to resize the disk.
   ```
   qemu-img resize FreeBSD-11.3-RELEASE-amd64.qcow2 +20G
   ```
   The filesystem should be automatically expanded when booting the first time.

4. Import and boot the image in virt-manager. Default user is root and there is no password.

5. Setup keyboard layout.
   ```
   kbdmap
   ```

6. Install packages bash and gdb. Clang and readelf should be installed already.
   The system may ask you to install `The package management tool`. Answer yes.
   ```
   pkg install bash gdb
   ```

7. Compile test program
   ```cpp
   #include <stdio.h>
   #include <stdlib.h>
    
   int main(int argc, char *argv[])
   {
       size_t *p = (size_t *) strtol(argv[1], NULL, 16);
       p[0] = 0xDEADBEEF;
       printf("RELRO: %p\n", p);
       return 0;
   }
   ```
   ```bash
   clang -g -Wl,-z,relro,-z,now test.c
   ```

8. Run checksec
   ```
   root@freebsd:~ # checksec --file=a.out
   RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      Symbols         FORTIFY Fortified       Fortifiable     FILE
   Partial RELRO   No canary found   NX enabled    No PIE          No RPATH   No RUNPATH   75) Symbols       No    1               2               a.out
   ```

9. Check that BIND_NOW is actually embedded

   Based on https://www.redhat.com/en/blog/hardening-elf-binaries-using-relocation-read-only-relro

   ```
   root@freebsd:~ # objdump -R a.out | grep -i printf
   0000000000600fe0 R_X86_64_JUMP_SLOT  printf
   ```
   ```
   root@freebsd:~ # gdb -q ./a.out
   (gdb) run 0000000000600fe0
   Starting program: /root/a.out 0000000000600fe0
   
   Program received signal SIGSEGV, Segmentation fault.
   0x000000000040080c in main (argc=2, argv=0x7fffffffeb08) at test.c:7
   7           p[0] = 0xDEADBEEF;
   Current language:  auto; currently minimal
   ```

## OS version and Kernel version

FreeBSD 11.3
```
root@freebsd:~ # uname -a
FreeBSD freebsd 11.3-RELEASE FreeBSD 11.3-RELEASE #0 r349754: Fri Jul  5 04:45:24 UTC 2019     root@releng2.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
```

| Software | Version |
| - | - |
| checksec | checksec v2.2.3, Brian Davis, github.com/slimm609/checksec.sh, Dec 2015 |
| clang    | FreeBSD clang version 8.0.0 (tags/RELEASE_800/final 356365) (based on LLVM 8.0.0) |
| gdb      | GNU gdb 6.1.1 [FreeBSD] |
| bash     | GNU bash, version 5.0.17(0)-release (amd64-portbld-freebsd11.3) |
| readelf  | readelf (elftoolchain r3614M) |

## Debug output

Run the same command as above to reproduce the error but include the --debug flag

```
root@freebsd:~ # checksec --debug --file=a.out
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH      Symbols         FORTIFY Fortified       Fortifiable     FILE
***function filecheck
***function filecheck->RELRO
Partial RELRO   
***function filecheck->canary
No canary found   
***function filecheck->nx
NX enabled    
***function filecheck->pie
No PIE          
***function filecheck->rpath
No RPATH   
***function filecheck->runpath
No RUNPATH   75) Symbols        ***function filecheck->fortify
No        1               2               a.out
```

**Comment by slimm609** [comment]
do you have a binary that you can share?  i should be able to analyze it without needing freebsd and will be much faster for me to debug.

**Comment by biboon** [comment]
Sure, there you go :
[a.out.zip](https://github.com/slimm609/checksec.sh/files/5217542/a.out.zip)
Interesting though, I just checked the binary with checksec on my Debian and the output is valid...

**Comment by slimm609** [comment]
https://github.com/slimm609/checksec.sh/tree/freebsd_relro

If you want to try this version.     freebsd readelf does not show bind_now but it still does have GNU_RELRO.    if it does have relro, then .got.plt is merged into .got with full RELRO.  So checking for .got.plt tells if its full or partial

**Comment by natpicone** [comment]
Hi, it looks like this branch is unavailable.
Has it been merged?

**Comment by natpicone** [comment]
> it has not been merged. I removed it when doing some cleanup. I never got confirmation on if it resolved the issue and I don't have freebsd to test. I repushed it up but will be closing in the next few weeks if I don't get confirmation https://github.com/slimm609/checksec.sh/tree/freebsd

Thanks a lot for pushing it back.
From my tests, it was working fine.
